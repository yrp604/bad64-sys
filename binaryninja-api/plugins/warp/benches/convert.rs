use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use criterion::{criterion_group, criterion_main, Criterion};
use std::path::PathBuf;
use warp_ninja::convert::from_bn_type;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 1] = ["atox.obj"];

pub fn type_conversion_benchmark(c: &mut Criterion) {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");

        // Bench function types.
        let functions: Vec<_> = view.functions().iter().map(|f| f.to_owned()).collect();
        c.bench_function("type conversion all functions", |b| {
            b.iter(|| {
                for func in &functions {
                    from_bn_type(&view, &func.function_type(), u8::MAX);
                }
            })
        });

        // Bench view types.
        let types = view.types().to_vec();
        c.bench_function("type conversion all types", |b| {
            b.iter(|| {
                for ty in &types {
                    from_bn_type(&view, &ty.ty, u8::MAX);
                }
            })
        });
    }
}

criterion_group!(benches, type_conversion_benchmark);
criterion_main!(benches);
