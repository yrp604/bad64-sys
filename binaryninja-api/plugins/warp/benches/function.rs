use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use criterion::{criterion_group, criterion_main, Criterion};
use rayon::prelude::*;
use std::path::PathBuf;
use warp_ninja::build_function;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 1] = ["atox.obj"];

pub fn function_benchmark(c: &mut Criterion) {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");

        let functions: Vec<_> = view.functions().iter().map(|f| f.to_owned()).collect();
        // Bench all functions sequentially
        c.bench_function("signature all functions", |b| {
            b.iter(|| {
                for func in &functions {
                    let _ = build_function(&func, || func.lifted_il().ok(), false);
                }
            })
        });

        // Bench all functions in parallel
        c.bench_function("signature all functions rayon", |b| {
            b.iter(|| {
                functions
                    .par_iter()
                    .filter_map(|func| {
                        Some(build_function(
                            func.as_ref(),
                            || func.lifted_il().ok(),
                            false,
                        ))
                    })
                    .collect::<Vec<_>>()
            })
        });
    }
}

criterion_group!(benches, function_benchmark);
criterion_main!(benches);
