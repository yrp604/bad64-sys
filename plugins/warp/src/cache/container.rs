use crate::container::Container;
use dashmap::DashMap;
use std::ops::Deref;
use std::sync::{Arc, OnceLock, RwLock};

pub static CONTAINER_CACHE: OnceLock<DashMap<String, Arc<RwLock<Box<dyn Container>>>>> =
    OnceLock::new();

pub fn for_cached_containers(f: impl Fn(&dyn Container)) {
    let containers_cache = CONTAINER_CACHE.get_or_init(Default::default);
    for container in containers_cache.iter() {
        if let Ok(guarded_container) = container.read() {
            f(guarded_container.deref().as_ref());
        }
    }
}

// TODO: The static lifetime here is a little wierd... (we need it to Box)
pub fn add_cached_container(container: impl Container + 'static) {
    let containers_cache = CONTAINER_CACHE.get_or_init(Default::default);
    let container_name = container.to_string();
    containers_cache.insert(container_name, Arc::new(RwLock::new(Box::new(container))));
}

pub fn cached_containers() -> Vec<Arc<RwLock<Box<dyn Container>>>> {
    let containers_cache = CONTAINER_CACHE.get_or_init(Default::default);
    containers_cache.iter().map(|c| c.clone()).collect()
}
