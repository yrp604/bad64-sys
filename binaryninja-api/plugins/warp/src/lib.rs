use crate::cache::{
    cached_adjacency_constraints, cached_call_site_constraints, cached_function_guid,
};
use crate::convert::{from_bn_symbol, from_bn_type};
use binaryninja::architecture::{
    Architecture, ImplicitRegisterExtend, Register as BNRegister, RegisterInfo,
};
use binaryninja::basic_block::BasicBlock as BNBasicBlock;
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::confidence::MAX_CONFIDENCE;
use binaryninja::function::{Function as BNFunction, NativeBlock};
use binaryninja::low_level_il::expression::{ExpressionHandler, LowLevelILExpressionKind};
use binaryninja::low_level_il::function::{
    FunctionMutability, LowLevelILFunction, NonSSA, RegularNonSSA,
};
use binaryninja::low_level_il::instruction::{
    InstructionHandler, LowLevelILInstruction, LowLevelILInstructionKind,
};
use binaryninja::low_level_il::{LowLevelILRegister, VisitorAction};
use binaryninja::rc::Ref as BNRef;
use std::path::PathBuf;
use warp::signature::basic_block::BasicBlockGUID;
use warp::signature::function::constraints::FunctionConstraints;
use warp::signature::function::{Function, FunctionGUID};

pub mod cache;
pub mod convert;
mod matcher;
/// Only used when compiled for cdylib target.
mod plugin;

pub fn core_signature_dir() -> PathBuf {
    // Get core signatures for the given platform
    let install_dir = binaryninja::install_directory();
    // macOS core dir is separate from the install dir.
    #[cfg(target_os = "macos")]
    let core_dir = install_dir.parent().unwrap().join("Resources");
    #[cfg(not(target_os = "macos"))]
    let core_dir = install_dir;
    core_dir.join("signatures")
}

pub fn user_signature_dir() -> PathBuf {
    binaryninja::user_directory().join("signatures/")
}

pub fn build_function<A: Architecture, M: FunctionMutability>(
    func: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> Function {
    let bn_fn_ty = func.function_type();
    Function {
        guid: cached_function_guid(func, llil),
        symbol: from_bn_symbol(&func.symbol()),
        ty: from_bn_type(&func.view(), &bn_fn_ty, MAX_CONFIDENCE),
        constraints: FunctionConstraints {
            // NOTE: Adding adjacent only works if analysis is complete.
            // NOTE: We do not filter out adjacent functions here.
            adjacent: cached_adjacency_constraints(func, |_| true),
            call_sites: cached_call_site_constraints(func),
            // TODO: Add caller sites (when adjacent and call sites are minimal)
            // NOTE: Adding caller sites only works if analysis is complete.
            caller_sites: Default::default(),
        },
    }
}

/// Basic blocks sorted from high to low.
pub fn sorted_basic_blocks(func: &BNFunction) -> Vec<BNRef<BNBasicBlock<NativeBlock>>> {
    let mut basic_blocks = func
        .basic_blocks()
        .iter()
        .map(|bb| bb.clone())
        .collect::<Vec<_>>();
    basic_blocks.sort_by_key(|f| f.start_index());
    basic_blocks
}

pub fn function_guid<A: Architecture, M: FunctionMutability>(
    func: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> FunctionGUID {
    let basic_blocks = sorted_basic_blocks(func);
    let basic_block_guids = basic_blocks
        .iter()
        .map(|bb| basic_block_guid(bb, llil))
        .collect::<Vec<_>>();
    FunctionGUID::from_basic_blocks(&basic_block_guids)
}

pub fn basic_block_guid<A: Architecture, M: FunctionMutability>(
    basic_block: &BNBasicBlock<NativeBlock>,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> BasicBlockGUID {
    let func = basic_block.function();
    let view = func.view();
    let arch = func.arch();
    let max_instr_len = arch.max_instr_len();

    // NOPs and useless moves are blacklisted to allow for hot-patchable functions.
    let is_blacklisted_instr = |instr: &LowLevelILInstruction<A, M, NonSSA<RegularNonSSA>>| {
        match instr.kind() {
            LowLevelILInstructionKind::Nop(_) => true,
            LowLevelILInstructionKind::SetReg(op) => {
                match op.source_expr().kind() {
                    LowLevelILExpressionKind::Reg(source_op)
                        if op.dest_reg() == source_op.source_reg() =>
                    {
                        match op.dest_reg() {
                            LowLevelILRegister::ArchReg(r) => {
                                // If this register has no implicit extend then we can safely assume it's a NOP.
                                // Ex. on x86_64 we don't want to remove `mov edi, edi` as it will zero the upper 32 bits.
                                // Ex. on x86 we do want to remove `mov edi, edi` as it will not have a side effect like above.
                                matches!(
                                    r.info().implicit_extend(),
                                    ImplicitRegisterExtend::NoExtend
                                )
                            }
                            LowLevelILRegister::Temp(_) => false,
                        }
                    }
                    _ => false,
                }
            }
            _ => false,
        }
    };

    let is_variant_instr = |instr: &LowLevelILInstruction<A, M, NonSSA<RegularNonSSA>>| {
        let is_variant_expr = |expr: &LowLevelILExpressionKind<A, M, NonSSA<RegularNonSSA>>| {
            // TODO: Checking the section here is slow, we should gather all section ranges outside of this.
            match expr {
                LowLevelILExpressionKind::ConstPtr(op)
                    if !view.sections_at(op.value()).is_empty() =>
                {
                    // Constant Pointer must be in a section for it to be relocatable.
                    // NOTE: We cannot utilize segments here as there will be a zero based segment.
                    true
                }
                LowLevelILExpressionKind::ExternPtr(_) => true,
                LowLevelILExpressionKind::Const(op) if !view.sections_at(op.value()).is_empty() => {
                    // Constant value must be in a section for it to be relocatable.
                    // NOTE: We cannot utilize segments here as there will be a zero based segment.
                    true
                }
                _ => false,
            }
        };

        // Visit instruction expressions looking for variant expression, [VisitorAction::Halt] means variant.
        instr.visit_tree(&mut |expr| {
            if is_variant_expr(&expr.kind()) {
                // Found a variant expression
                VisitorAction::Halt
            } else {
                VisitorAction::Descend
            }
        }) == VisitorAction::Halt
    };

    let basic_block_range = basic_block.start_index()..basic_block.end_index();
    let mut basic_block_bytes = Vec::with_capacity(basic_block_range.count());
    for instr_addr in basic_block.into_iter() {
        let mut instr_bytes = view.read_vec(instr_addr, max_instr_len);
        if let Some(instr_info) = arch.instruction_info(&instr_bytes, instr_addr) {
            instr_bytes.truncate(instr_info.length);
            if let Some(instr_llil) = llil.instruction_at(instr_addr) {
                // If instruction is blacklisted don't include the bytes.
                if !is_blacklisted_instr(&instr_llil) {
                    if is_variant_instr(&instr_llil) {
                        // Found a variant instruction, mask off entire instruction.
                        instr_bytes.fill(0);
                    }
                    // Add the instructions bytes to the basic blocks bytes
                    basic_block_bytes.extend(instr_bytes);
                }
            }
        }
    }

    BasicBlockGUID::from(basic_block_bytes.as_slice())
}

#[cfg(test)]
mod tests {
    use crate::cache::cached_function_guid;
    use binaryninja::binary_view::BinaryViewExt;
    use binaryninja::headless::Session;
    use std::path::PathBuf;
    use std::sync::OnceLock;

    static INIT: OnceLock<Session> = OnceLock::new();

    fn get_session<'a>() -> &'a Session {
        // TODO: This is not shared between other test modules, should still be fine (mutex in core now).
        INIT.get_or_init(|| Session::new().expect("Failed to initialize session"))
    }

    #[test]
    fn insta_signatures() {
        let session = get_session();
        let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
        for entry in std::fs::read_dir(out_dir).expect("Failed to read OUT_DIR") {
            let entry = entry.expect("Failed to read directory entry");
            let path = entry.path();
            if path.is_file() {
                let view = session.load(&path).expect("Failed to load view");
                let mut functions = view
                    .functions()
                    .iter()
                    .map(|f| cached_function_guid(&f, &f.low_level_il().unwrap()))
                    .collect::<Vec<_>>();
                functions.sort_by_key(|guid| guid.guid);
                let snapshot_name =
                    format!("snapshot_{}", path.file_stem().unwrap().to_string_lossy());
                insta::assert_debug_snapshot!(snapshot_name, functions);
            }
        }
    }
}
