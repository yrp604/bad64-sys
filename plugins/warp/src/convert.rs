pub mod symbol;
pub mod types;

use binaryninja::function::Comment as BNComment;
use binaryninja::function::Function as BNFunction;
use binaryninja::platform::Platform;
use binaryninja::rc::Ref;
use binaryninja::variable::{Variable as BNVariable, VariableSourceType};
pub use symbol::*;
pub use types::*;
use warp::r#type::class::function::{Location, RegisterLocation, StackLocation};
use warp::signature::comment::FunctionComment;
use warp::target::Target;

pub fn bn_var_to_location(bn_variable: BNVariable) -> Option<Location> {
    match bn_variable.ty {
        VariableSourceType::StackVariableSourceType => {
            let stack_loc = StackLocation {
                offset: bn_variable.storage,
            };
            Some(Location::Stack(stack_loc))
        }
        VariableSourceType::RegisterVariableSourceType => {
            let reg_loc = RegisterLocation {
                id: bn_variable.storage as u64,
            };
            Some(Location::Register(reg_loc))
        }
        VariableSourceType::FlagVariableSourceType => None,
    }
}

pub fn bn_comment_to_comment(func: &BNFunction, bn_comment: BNComment) -> FunctionComment {
    let offset = (bn_comment.addr as i64) - (func.start() as i64);
    FunctionComment {
        offset,
        text: bn_comment.comment,
    }
}

pub fn comment_to_bn_comment(func: &BNFunction, comment: FunctionComment) -> BNComment {
    BNComment {
        addr: comment
            .offset
            .checked_add_unsigned(func.start())
            .unwrap_or_default() as u64,
        comment: comment.text,
    }
}

pub fn platform_to_target(platform: &Platform) -> Target {
    let arch_name = platform.arch().name();
    let platform_name = platform.name();
    // We do not want to populate the platform if we are actually only the architecture.
    if arch_name == platform_name {
        Target {
            architecture: Some(arch_name),
            platform: None,
        }
    } else {
        Target {
            architecture: Some(arch_name),
            platform: Some(platform_name),
        }
    }
}

pub fn target_to_platform(target: Target) -> Option<Ref<Platform>> {
    // First try using the platform, then try using the arch.
    match Platform::by_name(&target.platform.unwrap()) {
        None => Platform::by_name(&target.architecture.unwrap()).map(|platform| platform),
        Some(platform) => Some(platform),
    }
}
