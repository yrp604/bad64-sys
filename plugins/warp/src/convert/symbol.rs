use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::rc::Ref as BNRef;
use binaryninja::symbol::Symbol as BNSymbol;
use binaryninja::symbol::SymbolType as BNSymbolType;
use warp::symbol::{Symbol, SymbolClass, SymbolModifiers};

pub fn from_bn_symbol(raw_symbol: &BNSymbol) -> Symbol {
    // TODO: Use this?
    let _is_export = raw_symbol.external();
    let raw_symbol_name = raw_symbol.raw_name();
    let symbol_name = raw_symbol_name.to_string_lossy();
    match raw_symbol.sym_type() {
        BNSymbolType::ImportAddress => {
            Symbol::new(
                symbol_name,
                SymbolClass::Function,
                // TODO: External = symbolic i guess
                SymbolModifiers::External,
            )
        }
        BNSymbolType::Data => {
            Symbol::new(
                symbol_name,
                // TODO: Data?
                SymbolClass::Data,
                SymbolModifiers::default(),
            )
        }
        BNSymbolType::Symbolic => {
            Symbol::new(
                symbol_name,
                SymbolClass::Function,
                // TODO: External = symbolic i guess
                SymbolModifiers::External,
            )
        }
        BNSymbolType::LocalLabel => {
            // TODO: This is a placeholder for another symbol.
            Symbol::new(symbol_name, SymbolClass::Data, SymbolModifiers::External)
        }
        BNSymbolType::External => Symbol::new(
            symbol_name,
            // TODO: External data?
            SymbolClass::Function,
            SymbolModifiers::External,
        ),
        BNSymbolType::ImportedData => {
            Symbol::new(symbol_name, SymbolClass::Data, SymbolModifiers::External)
        }
        BNSymbolType::LibraryFunction | BNSymbolType::Function => Symbol::new(
            symbol_name,
            SymbolClass::Function,
            SymbolModifiers::default(),
        ),
        BNSymbolType::ImportedFunction => Symbol::new(
            symbol_name,
            SymbolClass::Function,
            // TODO: Exported?
            SymbolModifiers::External,
        ),
    }
}

pub fn to_bn_symbol_at_address(view: &BinaryView, symbol: &Symbol, addr: u64) -> BNRef<BNSymbol> {
    let is_external = symbol.modifiers.contains(SymbolModifiers::External);
    let _is_exported = symbol.modifiers.contains(SymbolModifiers::Exported);
    let symbol_type = match symbol.class {
        SymbolClass::Function if is_external => BNSymbolType::ImportedFunction,
        // TODO: We should instead make it a Function, however due to the nature of the imports we are setting them to library for now.
        SymbolClass::Function => BNSymbolType::LibraryFunction,
        SymbolClass::Data if is_external => BNSymbolType::ImportedData,
        SymbolClass::Data => BNSymbolType::Data,
        _ => BNSymbolType::Data,
    };
    let raw_name = symbol.name.as_str();
    let mut symbol_builder = BNSymbol::builder(symbol_type, &symbol.name, addr);
    // Demangle symbol name (short is with simplifications).
    if let Some(arch) = view.default_arch() {
        if let Some((full_name, _)) =
            binaryninja::demangle::demangle_generic(&arch, raw_name, Some(view), false)
        {
            symbol_builder = symbol_builder.full_name(full_name);
        }
        if let Some((short_name, _)) =
            binaryninja::demangle::demangle_generic(&arch, raw_name, Some(view), false)
        {
            symbol_builder = symbol_builder.short_name(short_name);
        }
    }
    symbol_builder.create()
}
