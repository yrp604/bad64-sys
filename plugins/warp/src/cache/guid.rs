use crate::cache::FunctionID;
use crate::convert::from_bn_symbol;
use crate::function_guid;
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::Function as BNFunction;
use binaryninja::low_level_il::function::{FunctionMutability, LowLevelILFunction, NonSSA};
use binaryninja::rc::Ref as BNRef;
use binaryninja::symbol::Symbol as BNSymbol;
use std::collections::HashSet;
use uuid::Uuid;
use warp::signature::constraint::Constraint;
use warp::signature::function::FunctionGUID;

/// Try to get the cached function GUID from the metadata.
///
/// If not cached, we will use `lifted_il_accessor` to retrieve the lifted IL and create the GUID.
///
/// `lifted_il_accessor` exists as it is to allow the retrieval of a cached GUID without incurring
/// the cost of building the IL (if it no longer exists).
pub fn cached_function_guid<M: FunctionMutability>(
    function: &BNFunction,
    lifted_il_accessor: impl Fn() -> Option<BNRef<LowLevelILFunction<M, NonSSA>>>,
) -> Option<FunctionGUID> {
    let cached_guid = try_cached_function_guid(function);
    if let Some(cached_guid) = cached_guid {
        return Some(cached_guid);
    }

    let function_guid = function_guid(function, lifted_il_accessor()?.as_ref());
    function.store_metadata(
        "warp_function_guid",
        &function_guid.as_bytes().to_vec(),
        false,
    );
    Some(function_guid)
}

pub fn try_cached_function_guid(function: &BNFunction) -> Option<FunctionGUID> {
    let metadata = function.query_metadata("warp_function_guid")?;
    let raw_metadata = metadata.get_raw()?;
    let uuid = Uuid::from_slice(raw_metadata.as_slice()).ok()?;
    Some(FunctionGUID::from(uuid))
}

pub fn cached_constraints<F>(function: &BNFunction, filter: F) -> HashSet<Constraint>
where
    F: Fn(&BNFunction) -> bool,
{
    // TODO: Implied constraints, symbol name, image offset
    let cs_constraints = cached_call_site_constraints(function);
    let adj_constraints = cached_adjacency_constraints(function, filter);
    cs_constraints.union(&adj_constraints).cloned().collect()
}

pub fn cached_call_site_constraints(function: &BNFunction) -> HashSet<Constraint> {
    let cache = ConstraintBuilder;
    cache.call_site_constraints(function)
}

pub fn cached_adjacency_constraints<F>(function: &BNFunction, filter: F) -> HashSet<Constraint>
where
    F: Fn(&BNFunction) -> bool,
{
    let cache = ConstraintBuilder;
    cache.adjacency_constraints(function, filter)
}

#[derive(Clone, Debug, Default)]
pub struct ConstraintBuilder;

impl ConstraintBuilder {
    pub fn call_site_constraints(&self, function: &BNFunction) -> HashSet<Constraint> {
        let view = function.view();
        let func_id = FunctionID::from(function);
        let func_start = function.start();
        let func_platform = function.platform();
        let mut constraints = HashSet::new();
        for call_site in &function.call_sites() {
            for cs_ref_addr in view.code_refs_from_addr(call_site.address, Some(function)) {
                match view.function_at(&func_platform, cs_ref_addr) {
                    Some(cs_ref_func) => {
                        // Call site is a function, constrain on it.
                        let cs_ref_func_id = FunctionID::from(cs_ref_func.as_ref());
                        if cs_ref_func_id != func_id {
                            let call_site_offset: i64 =
                                call_site.address.wrapping_sub(func_start) as i64;
                            // TODO: If the function is thunk we should also insert the called function.
                            constraints.extend(
                                self.related_function_constraint(&cs_ref_func, call_site_offset),
                            );
                        }
                    }
                    None => {
                        // We could be dealing with an extern symbol, get the symbol as a constraint.
                        let call_site_offset: i64 =
                            call_site.address.wrapping_sub(func_start) as i64;
                        if let Some(call_site_sym) = view.symbol_by_address(cs_ref_addr) {
                            constraints.insert(
                                self.related_symbol_constraint(&call_site_sym, call_site_offset),
                            );
                        }
                    }
                }
            }
        }
        constraints
    }

    pub fn adjacency_constraints<F>(&self, function: &BNFunction, filter: F) -> HashSet<Constraint>
    where
        F: Fn(&BNFunction) -> bool,
    {
        let view = function.view();
        let func_id = FunctionID::from(function);
        let func_start = function.start();
        let mut constraints = HashSet::new();

        let mut func_addr_constraint = |func_start_addr| {
            // NOTE: We could potentially have dozens of functions all at the same start address.
            for curr_func in &view.functions_at(func_start_addr) {
                let curr_func_id = FunctionID::from(curr_func.as_ref());
                if curr_func_id != func_id && filter(curr_func.as_ref()) {
                    // NOTE: For this to work the GUID has to have already been cached. If not it will just be the symbol.
                    // Function adjacent to another function, constrain on the pattern.
                    let curr_addr_offset = (func_start_addr as i64) - func_start as i64;
                    constraints
                        .extend(self.related_function_constraint(&curr_func, curr_addr_offset));
                }
            }
        };

        let mut before_func_start = func_start;
        for _ in 0..2 {
            before_func_start = view.function_start_before(before_func_start);
            func_addr_constraint(before_func_start);
        }

        let mut after_func_start = func_start;
        for _ in 0..2 {
            after_func_start = view.function_start_after(after_func_start);
            func_addr_constraint(after_func_start);
        }

        constraints
    }

    /// Construct a function constraint, must pass the offset at which it is located.
    pub fn related_function_constraint(
        &self,
        function: &BNFunction,
        offset: i64,
    ) -> Vec<Constraint> {
        let mut constraints = vec![];
        if let Some(guid) = try_cached_function_guid(function) {
            let guid_constraint = Constraint::from_function(&guid, Some(offset));
            constraints.push(guid_constraint);
        }
        let symbol_constraint = self.related_symbol_constraint(&function.symbol(), offset);
        constraints.push(symbol_constraint);
        constraints
    }

    /// Construct a symbol constraint, must pass the offset at which it is located.
    pub fn related_symbol_constraint(&self, symbol: &BNSymbol, offset: i64) -> Constraint {
        let mut symbol = from_bn_symbol(symbol);
        symbol.name = clean_symbol_name(&symbol.name);
        Constraint::from_symbol(&symbol, Some(offset))
    }
}

/// Cleans various internal symbol prefixes and suffixes for consistency.
///
/// This is very important for getting matching symbol constraints.
///
/// Examples:
/// - "__imp__RemoveDirectoryW@4" -> "RemoveDirectoryW"
/// - "__free_base" -> "free_base"
/// - "__impl__free_base" -> "free_base"
/// - "j___free_base" -> "free_base"
/// - "j_free_base" -> "free_base"
/// - "_free_base" -> "free_base"
pub fn clean_symbol_name(symbol_name: &str) -> String {
    // Handle MSVC-style imported symbols
    let without_imp = symbol_name.strip_prefix("__imp__").unwrap_or(symbol_name);

    // Handle jump thunk prefix
    let without_jump = without_imp.strip_prefix("j_").unwrap_or(without_imp);

    // Strip all remaining leading underscores
    let mut result = without_jump;
    while result.starts_with('_') {
        result = &result[1..];
    }

    // Remove stdcall decoration (@N suffix)
    match result.find('@') {
        Some(pos) => result[..pos].to_string(),
        None => result.to_string(),
    }
}
