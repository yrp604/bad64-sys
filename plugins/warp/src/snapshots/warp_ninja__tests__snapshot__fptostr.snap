---
source: plugins/warp/src/lib.rs
expression: functions
---
[
    FunctionGUID {
        guid: 9d4dba66-7106-5215-8e64-a47279f67dfe,
    },
]
